### PR TITLE
feat: add determine single namespace item num limit logic

### PR DESCRIPTION
**Pull request by youngzil**
## What's the purpose of this PR

Added a new parameter `item.num.limit` ，the purpose is to limit the number of configuration items in the same namespace

In actual use, when modifying configurations in batches, if the number of configurations exceeds a certain value, it is easy to trigger a timeout or page crash.

## Which issue(s) this PR fixes:
Fixes #

## Brief changelog

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit tests to verify the code.
- [ ] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [ ] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo/blob/master/CHANGES.md).

## Summary by CodeRabbit

- **New Features**
  - Introduced a limit on the number of items that can be created in a namespace, enhancing data integrity.
  - Added functionality to count non-empty items in a specified namespace.
  - New system parameters for enabling item limits and setting maximum item counts.
  - Enhanced configuration management with new features for security and gray release strategies.
  - Administrators can now globally search for values within the system.

- **Bug Fixes**
  - Implemented validation checks to prevent exceeding item limits, improving error handling.
  - Resolved issues with duplicate comments and missing items in published namespaces.
  
- **Documentation**
  - Updated user guides to include instructions for enabling and configuring item limits.
  - Expanded documentation with details on gray release functionality and security best practices.

### CODE REVIEW

**Comment by nobodyiam** [comment]
Duplicated with #5228

File: CHANGES.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ Apollo 2.4.0
 * [Feature support portal restTemplate Client connection pool config](https://github.com/apolloconfig/apollo/pull/5200)
 * [Feature added the ability for administrators to globally search for Value](https://github.com/apolloconfig/apollo/pull/5182)
 * [Fix link namespace published items show missing some items](https://github.com/apolloconfig/apollo/pull/5240)
+* [Feature added limit the number of items under a single namespace](https://github.com/apolloconfig/apollo/pull/5228)
```
Comment (nobodyiam) [inline CHANGES.md:12]:
```suggestion
* [Feature add limit for items count per namespace](https://github.com/apolloconfig/apollo/pull/5227)
```

File: apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/ItemController.java [hunk 2]
```diff
@@ -78,6 +80,14 @@
     if (managedEntity != null) {
       throw BadRequestException.itemAlreadyExists(entity.getKey());
     }
+
+    if (bizConfig.isItemNumLimitEnabled()) {
+      int itemCount = itemService.findNonEmptyItemCount(entity.getNamespaceId());
+      if (itemCount >= bizConfig.itemNumLimit()) {
+        throw new BadRequestException("current namespace item count=[" + itemCount + "], single namespace max allow count=[ " + bizConfig.itemNumLimit() + "]");
```
[the remark this answers is not in the document]
Comment (nobodyiam) [reply apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/ItemController.java:87]: This is a valid suggestion.

File: apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/ItemController.java [hunk 3]
```diff
@@ -78,6 +80,14 @@ public ItemDTO create(@PathVariable("appId") String appId,
...
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (nobodyiam) [reply apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/ItemController.java:90]: This is an issue, but I believe it is more of a soft limit than a hard one, so it should be fine.

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/config/BizConfig.java [hunk 4]
```diff
@@ -99,6 +102,15 @@ public int itemValueLengthLimit() {
     return checkInt(limit, 5, Integer.MAX_VALUE, DEFAULT_ITEM_VALUE_LENGTH);
   }
 
+  public boolean isItemNumLimitEnabled() {
+    return getBooleanProperty("item.num.limit.enabled", false);
+  }
+
+  public int itemNumLimit() {
+    int limit = getIntProperty("item.num.limit", DEFAULT_MAX_ITEM_NUM);
+    return checkInt(limit, 5, Integer.MAX_VALUE, DEFAULT_MAX_ITEM_NUM);
```
Comment (nobodyiam) [inline apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/config/BizConfig.java:111]: Why is the minimum 5?

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/repository/ItemRepository.java [hunk 5]
```diff
@@ -64,4 +64,7 @@ public interface ItemRepository extends PagingAndSortingRepository<Item, Long> {
   @Query("update Item set IsDeleted = true, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000), DataChange_LastModifiedBy = ?2 where NamespaceId = ?1 and IsDeleted = false")
   int deleteByNamespaceId(long namespaceId, String operator);
 
+  @Query("select count(*) from Item where namespaceId = :namespaceId and key <>''")
+  int countByNamespaceIdAndFilterKeyEmpty(@Param("namespaceId") long namespaceId);
```
Comment (nobodyiam) [inline apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/repository/ItemRepository.java:68]: Is the following statement the same?
```suggestion
  int countByNamespaceIdAndKeyIsNotEmpty(long namespaceId);
```

**Review by nobodyiam — CHANGES_REQUESTED** [review]

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/repository/ItemRepository.java (hunk 5 above)
Comment (youngzil) [reply apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/repository/ItemRepository.java:68]: `countByNamespaceIdAndKeyIsNotEmpty` or `countByNamespaceIdAndKeyIsNotNullAndKeyNotEmpty` will result in the following exception:

> IsEmpty / IsNotEmpty can only be used on collection properties!

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/config/BizConfig.java (hunk 4 above)
Comment (youngzil) [reply apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/config/BizConfig.java:111]: The minimum value of 5 is set based on the values ​​of `item.key.length.limit `and `item.value.length.limit`. Considering that actual use will not set a value smaller than 5.
Can you give me a suggested minimum?

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ItemService.java [hunk 6]
```diff
@@ -135,6 +135,10 @@ public List<Item> findItemsModifiedAfterDate(long namespaceId, Date date) {
     return itemRepository.findByNamespaceIdAndDataChangeLastModifiedTimeGreaterThan(namespaceId, date);
   }
 
+  public int findNonEmptyItemCount(long namespaceId) {
+    return itemRepository.countByNamespaceIdAndFilterKeyEmpty(namespaceId);
+  }
```
[the remark this answers is not in the document]
Comment (youngzil) [reply apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ItemService.java:140]: Batch modification will result in a large number of queries to the DB

**Review by nobodyiam — DISMISSED** [review]
LGTM

**Comment by nobodyiam** [comment]
Would you please fix the unit test error?

**Comment by youngzil** [comment]
> Would you please fix the unit test error?

@nobodyiam  I have fixed this unit test, please review it again

**Pull request merged**